cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -97,28 +100,154 @@ func recoverySystemEssentialSnaps(seedDir, recoverySystem string, essentialTypes
+	if err != nil {
+		return err
+	}
+	if !allMounted {
+		return nil
+	}
+
+	// n+1: final step: write $(ubuntu_data)/var/lib/snapd/modeenv - this
+	//      is the tmpfs we just created above
+	modeEnv := &boot.Modeenv{
+		Mode:           "install",
+		RecoverySystem: recoverySystem,
+	}
+	if err := modeEnv.WriteTo(boot.InitramfsWritableDir); err != nil {
+		return err
+	}
+	// and disable cloud-init in install mode
+	if err := sysconfig.DisableCloudInit(boot.InitramfsWritableDir); err != nil {
+		return err
+	}
+
+	// n+3: done, no output, no error indicates to initramfs we are done
+	//      with mounting stuff
+	return nil
+}
+
+// copyUbuntuDataAuth copies the authenication files like
+//  - extrausers passwd,shadow etc
+//  - sshd host configuration
+//  - user .ssh dir
+// to the target directory. This is used to copy the authentication
+// data from a real uc20 ubuntu-data partition into a ephemeral one.
+func copyUbuntuDataAuth(src, dst string) error {
+	for _, globEx := range []string{
+		"system-data/var/lib/extrausers/*",
+		"system-data/etc/ssh/*",
+		"user-data/*/.ssh/*",
+		// this ensures we also get non-ssh enabled accounts copied
+		"user-data/*/.profile",
+	} {
+		matches, err := filepath.Glob(filepath.Join(src, globEx))
+		if err != nil {
+			return err
+		}
+		for _, p := range matches {
+			comps := strings.Split(strings.TrimPrefix(p, src), "/")
+			for i := range comps {
+				part := filepath.Join(comps[0 : i+1]...)
+				fi, err := os.Stat(filepath.Join(src, part))
+				if err != nil {
+					return err
+				}
+				if fi.IsDir() {
+					if err := os.Mkdir(filepath.Join(dst, part), fi.Mode()); err != nil && !os.IsExist(err) {
+						return err
+					}
+					st, ok := fi.Sys().(*syscall.Stat_t)
+					if !ok {
+						return fmt.Errorf("cannot get stat data: %v", err)
+					}
+					if err := os.Chown(filepath.Join(dst, part), int(st.Uid), int(st.Gid)); err != nil {
+						return err
+					}
+				} else {
+					if err := osutil.CopyFile(p, filepath.Join(dst, part), osutil.CopyFlagPreserveAll); err != nil {
+						return err
+					}
+				}
+			}
+		}
+	}
+
+	// ensure the user state is transferred as well
+	if err := state.CopyState(filepath.Join(src, "system-data/var/lib/snapd/state.json"), filepath.Join(dst, "system-data/var/lib/snapd/state.json"), []string{"auth.users"}); err != nil && err != state.ErrNoState {
+		return fmt.Errorf("cannot copy user state: %v", err)
+	}
+
+	return nil
+}
+
+func generateMountsModeRecover(recoverySystem string) error {
+	allMounted, err := generateMountsCommonInstallRecover(recoverySystem)
+	if err != nil {
+		return err
+	}
+	if !allMounted {
+		return nil
+	}
+
+	// n+1: mount ubuntu-data for recovery
+	isRecoverDataMounted, err := osutilIsMounted(boot.InitramfsRecoverUbuntuDataDir)
+	if err != nil {
+		return err
+	}
+	if !isRecoverDataMounted {
+		fmt.Fprintf(stdout, "/dev/disk/by-label/ubuntu-data %s\n", boot.InitramfsRecoverUbuntuDataDir)
```

Comment:
I'm not sure how we actually want to do this mount in the end, because it might fail but we might still want to get the user to a shell without access to data in some cases. It's orthogonal to getting the ephemeral system running.